### PR TITLE
Properly check calls to `new` on `T::Class[...]` types

### LINKED ISSUE #7274: Missing error calling `new` in instance method on `Class`
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20Class%0A%20%20def%20my_new%0A%20%20%20%20%23%20There%20should%20be%20an%20error%20here%2C%20but%20Sorbet%20doesn't%20currently%20model%20that.%0A%20%20%20%20%23%20If%20we%20made%20the%20Class_new%20intrinsic%20dispatch%20to%20%60initialize%60%20on%0A%20%20%20%20%23%20%60%3CAttachedClass%3E%60%20we%20can%20_almost_%20get%20the%20behavior%20we%20want%2C%20but%20doing%0A%20%20%20%20%23%20that%20would%20require%20%60Class%3A%3A%3CAttachedClass%3E%60%20to%20be%20%60upper%3A%20BasicObject%60%0A%20%20%20%20%23%20which%20is%20inconvenient.%0A%20%20%20%20%23%0A%20%20%20%20%23%20So%20we%20might%20want%20to%20do%20some%20ad-hoc%20defaulting%20in%20the%20Class_new%20intrinsic%0A%20%20%20%20%23%20to%20either%20dispatch%20to%20%60%3CAttachedClass%3E%60%2C%20or%20if%20it's%20%60%3Ctop%3E%60%2C%20then%0A%20%20%20%20%23%20dispatch%20to%20%60BasicObject%60.%0A%20%20%20%20%23%0A%20%20%20%20%23%20Note%3A%20normally%20we%20don't%20do%20these%20sorts%20of%20%3Ctop%3E%20-%3E%20BasicObject%0A%20%20%20%20%23%20conversions%2C%20because%20of%20parametricity%2C%20so%20this%20would%20be%20a%20little%20bit%0A%20%20%20%20%23%20frowned%20upon%20but%20maybe%20worth%20it.%0A%20%20%20%20%23%0A%20%20%20%20%23%20https%3A%2F%2Fblog.jez.io%2Fsorbet-parametricity%2F%0A%20%20%20%20new%281%29%0A%20%20end%0Aend%0A%0Aclass%20A%0Aend%0A%0AA.my_new%20%23%20in%20%60initialize'%3A%20wrong%20number%20of%20arguments%20%28given%201%2C%20expected%200%29%20%28ArgumentError%29)

```ruby
# typed: true

class Class
  def my_new
    # There should be an error here, but Sorbet doesn't currently model that.
    # If we made the Class_new intrinsic dispatch to `initialize` on
    # `<AttachedClass>` we can _almost_ get the behavior we want, but doing
    # that would require `Class::<AttachedClass>` to be `upper: BasicObject`
    # which is inconvenient.
    #
    # So we might want to do some ad-hoc defaulting in the Class_new intrinsic
    # to either dispatch to `<AttachedClass>`, or if it's `<top>`, then
    # dispatch to `BasicObject`.
    #
    # Note: normally we don't do these sorts of <top> -> BasicObject
    # conversions, because of parametricity, so this would be a little bit
    # frowned upon but maybe worth it.
    #
    # https://blog.jez.io/sorbet-parametricity/
    new(1)
  end
end

class A
end

A.my_new # in `initialize': wrong number of arguments (given 1, expected 0) (ArgumentError)
```

#### Observed output

```
No errors! Great job.
```

#### Expected behavior

See the comment in the test.

It's unclear whether this is common enough to warrant building. I just happened to notice it while reading the code.

- - -

### PR DESCRIPTION
**Pull request by jez**
### Motivation

Fixes #7274

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_OZCyBKxUKRJW2h
→ https://go/builds/bui_OZCy1vxOV4SvgH
→ https://go/builds/bui_OZCyQvlX388ove

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_OZDVJ6BDdZ5JwJ
→ https://go/builds/bui_OZDVZDJEy71zHb
→ https://go/builds/bui_OZDVQfWanoTAIw

**Comment by jez** [comment]
This produces only 15 errors on Stripe's codebase, and all of them look like they're reasonable type errors—if you're trying to call `new` on a class which can be subclassed, then your `new` method implicitly has some API. If you're getting the dispatch going all the way up to `BasicObject#initialize` but you're passing it some arguments, then your `T::Class[...]` type is wrong.

There's still a problem where we don't check override compatibility on any `initialize` methods at runtime unless the parent is declared with `abstract` or `override` ([see here](https://github.com/sorbet/sorbet/pull/6888/files#diff-b565686f17b8592d5c1e544cd639aaac3244869ac059f2db416a2ac24aa94675R9)) but that's a problem for another day.

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_OcvLXlhjrl1Sze
→ https://go/builds/bui_OcvLueY4yqbFEz
→ https://go/builds/bui_OcvLA2QZDSqBUv

File: core/types/calls.cc [hunk 1]
```diff
@@ -1995,21 +1995,27 @@ class Class_new : public IntrinsicMethod {
...
-
-        auto attachedClass = self.data(gs)->attachedClass(gs);
-        if (!attachedClass.exists()) {
-            // If someone takes `klass: T::Class[T.anything]` and calls `klass.new`, the call is
-            // actually going to be on an "instance" not a singleton (Class.new, the one on the
-            // singleton, is the one that defines a new class at runtime).
-            //
-            // In that case, there's no attachedClass to look for an `initialize` method on.
-            // We could _maybe_ imagine trying to dispatch to `initialize` on the `<AttachedClass>`
-            // type argument? But I haven't thought about what the consequences of that would be.
```
Comment (jez) [inline core/types/calls.cc:2009]: I've now thought through the consequences, and I think it's the right thing to do 🙂

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_Od03LzXK4DnREr
→ https://go/builds/bui_Od03o8zXsCdJnY
→ https://go/builds/bui_Od035zY18xRf8v

**Comment by jez** [comment]
Converting back to a draft, for two reasons.

### 1. #1317, a bug with abstract methods, also applies to `initialize`:

#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20Module%3B%20include%20T%3A%3ASig%3B%20end%0A%0Amodule%20MyInterface%0A%20%20extend%20T%3A%3AHelpers%0A%20%20abstract!%0A%0A%20%20sig%20%7Babstract.params%28x%3A%20Integer%29.void%7D%0A%20%20def%20initialize%28x%29%3B%20end%0Aend%0A%0Aclass%20A%0A%20%20include%20MyInterface%0Aend)

```ruby
# typed: true
class Module; include T::Sig; end

module MyInterface
  extend T::Helpers
  abstract!

  sig {abstract.params(x: Integer).void}
  def initialize(x); end
end

class A
  include MyInterface
end
```

#### Observed output

```
No errors! Great job.
```

It seems painful to require people to go annotate all their interfaces with abstract `initialize` methods if Sorbet isn't even going to check that those `initialize` methods are implemented with the correct arity.

### 2. #7309, a bug with the inheritance chain of modules

#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20Module%3B%20include%20T%3A%3ASig%3B%20end%0A%0Amodule%20MyInterface%0A%20%20def%20example%0A%20%20%20%20initialize%0A%0A%20%20%20%20T.let%28self%2C%20BasicObject%29%0A%20%20end%0Aend%0A%0Asig%20%7Bparams%28klass%3A%20T%3A%3AClass%5BMyInterface%5D%29.void%7D%0Adef%20example%28klass%29%0A%20%20klass.new%0Aend)

```ruby
# typed: true
class Module; include T::Sig; end

module MyInterface
  def example
    initialize

    T.let(self, BasicObject)
  end
end

sig {params(klass: T::Class[MyInterface]).void}
def example(klass)
  klass.new
end
```

#### Observed output

```
No errors! Great job.
```

Again, it would be painful to make people go specify an abstract `initialize` but then have that `initialize` magically appear.

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_OdjipLJ4v7WvHL
→ https://go/builds/bui_OdjiVKevcldBBP
→ https://go/builds/bui_OdjjEqdNnIr32l

**Review by neilparikh — APPROVED** [review]

**Pull request merged**